crates/ide-assists/src/handlers/introduce_named_generic.rs

```diff
@@ -25,42 +25,49 @@ pub(crate) fn introduce_named_generic(acc: &mut Assists, ctx: &AssistContext<'_>
 
     let type_bound_list = impl_trait_type.type_bound_list()?;
 
+    // FIXME: Is this node appropriate to use for SyntaxEditor in this case?
+    let parent_node = match ctx.covering_element() {
+        SyntaxElement::Node(n) => n,
+        SyntaxElement::Token(t) => t.parent()?,
+    };
+    let make = SyntaxFactory::new();
+
     let target = fn_.syntax().text_range();
     acc.add(
         AssistId("introduce_named_generic", AssistKind::RefactorRewrite),
         "Replace impl trait with generic",
         target,
         |edit| {
-            let impl_trait_type = edit.make_mut(impl_trait_type);
-            let fn_ = edit.make_mut(fn_);
-            let fn_generic_param_list = fn_.get_or_create_generic_param_list();
-
-            let existing_names = fn_generic_param_list
-                .generic_params()
-                .flat_map(|param| match param {
-                    ast::GenericParam::TypeParam(t) => t.name().map(|name| name.to_string()),
-                    p => Some(p.to_string()),
-                })
-                .collect_vec();
+            let mut editor = edit.make_editor(&parent_node);
+
+            let existing_names = match fn_.generic_param_list() {
+                Some(generic_param_list) => generic_param_list
+                    .generic_params()
+                    .flat_map(|param| match param {
+                        ast::GenericParam::TypeParam(t) => t.name().map(|name| name.to_string()),
+                        p => Some(p.to_string()),
+                    })
+                    .collect_vec(),
+                None => Vec::new(),
+            };
             let type_param_name = suggest_name::NameGenerator::new_with_names(
                 existing_names.iter().map(|s| s.as_str()),
             )
             .for_impl_trait_as_generic(&impl_trait_type);
 
-            let type_param = make::type_param(make::name(&type_param_name), Some(type_bound_list))
-                .clone_for_update();
-            let new_ty = make::ty(&type_param_name).clone_for_update();
+            let type_param = make.type_param(make.name(&type_param_name), Some(type_bound_list));
+            let new_ty = make.ty(&type_param_name);
 
-            ted::replace(impl_trait_type.syntax(), new_ty.syntax());
-            fn_generic_param_list.add_generic_param(type_param.into());
+            editor.replace(impl_trait_type.syntax(), new_ty.syntax());
+            let generic_param = syntax::ast::GenericParam::from(type_param);
+            fn_.syntax_editor_add_generic_param(&mut editor, generic_param.clone());
```

Comment:
```suggestion
            fn_.syntax_editor_add_generic_param(&mut editor, type_param.clone().into());
```

Reply:
Then we would need to clone `type_param` on the previous line, so it would be:
```rust
let generic_param = syntax::ast::GenericParam::from(type_param.clone());
fn_.syntax_editor_add_generic_param(&mut editor, type_param.clone().into());
```
Do you prefer this over the current code?

Reply:
I think we can just remove `generic_param` and replace with `type_param` below

Reply:
We actually use it later to add annotation

Reply:
Right, but can we add the annotation using `type_param` instead?

Reply:
We can. should be fixed now